File: years/2022/clojure/src/day_04.clj
(ns day-04
  (:gen-class))

(require '[clojure.string :as str]
         '[clojure.set :as sets])

(defn read-file [filename]
  (str/split-lines
   (slurp filename)))

(defn make-section [s]
  (let [[start end] (map #(Integer/parseInt %) (str/split s #"-"))]
    (set (range start (inc end)))))

(defn sections [s]
  (map make-section (str/split s #",")))

(defn fully-contained? [s1 s2]
  (or (sets/subset? s1 s2) (sets/subset? s2 s1)))

(defn overlap? [s1 s2]
  (seq (sets/intersection s1 s2)))

(defn part1 [filename]
  (count (filter #(apply fully-contained? (sections %)) (read-file filename))))

(defn part2 [filename]
  (count (filter #(apply overlap? (sections %)) (read-file filename))))

(defn -main []
  (printf "day 04 part 1: %d%n" (part1 "../data/04.txt"))
  (printf "day 04 part 2: %d%n" (part2 "../data/04.txt")))

File: years/2016/elixir/lib/aoc/day_04.ex
defmodule AoC.Day04 do
  @moduledoc false

  @spec run() :: :ok
  def run do
    IO.puts("day 04 part 1: #{AoC.Day04.run_part_1("../data/04.txt")}")
    IO.puts("day 04 part 2: #{AoC.Day04.run_part_2("../data/04.txt")}")
  end

  @spec run_part_1(String.t()) :: number()
  def run_part_1(filename) do
    filename
    |> File.stream!()
    |> part_1()
  end

  @spec run_part_2(String.t()) :: number()
  def run_part_2(filename) do
    filename
    |> File.stream!()
    |> part_2()
  end

  @spec part_1(Enumerable.t()) :: integer()
  def part_1(data) do
    data
    |> Enum.count()
  end

  @spec part_2(Enumerable.t()) :: integer()
  def part_2(data) do
    data
    |> Enum.count()
  end
end
